drivers: pwm_nrf5_sw: Fix the driver and make it support nRF9x SoCs

anangl opened the pull request:
This PR makes the blink_led sample working (again) on boards with nRF52 series SoCs. The SW PWM driver is selected for them instead of the default HW one, as the latter cannot handle periods longer than ~262 ms (this is a hardware limitation).
It also fixes a couple of issues (details below) found in the pwm_nrf5_sw driver and makes it support nRF9x SoCs as well, so that the blink_led sample can be run on boards equipped with those.

**drivers: pwm_nrf5_sw: Use dynamically adjusted timer prescaler**

Adjust the timer prescaler used by the pwm_nrf5_sw driver according to
the requested period length, so that it does not need to be specified
at the build time. Remove the incorrect "scaling" that was done to the
requested pulse and period cycles to fit them into 16 bits.

**samples: blink_led: Use nRF SW PWM driver for nRF52 boards**

The nrfx Hardware PWM driver cannot be used in this sample
since the driver does not allow to set a period longer than ~262 ms.

**drivers: pwm_nrf5_sw: Correct the use of GPIOTE channels and TIMER**

GPIOTE base channel number taken from DeviceTree was not used
consistenly, hence the driver could properly work only when this
number had the value of 0.
TIMER was not stopped during the reconfiguration of the channel
and this could lead to the inversion of the PWM polarity.
The config structure was removed on the occasion as it unnecessarily
complicated the code of this currently single-instance only driver.

**drivers: pwm_nrf5_sw: Use nrfx HALs instead of direct register accesses**

Now instead of accessing peripheral registers directly, the driver
uses the nrfx HAL functions. This makes the code easier to maintain
and also gives the possibility to use pins from GPIO port P1 on
nRF52840 (absolute numbers 32-47) as the GPIO HAL provides the needed
translation.

**drivers: pwm_nrf5_sw: Refactor the code**

Move all channel initialization code that does not need to be executed
on every pin reconfiguration (like initialization of the PPI channels)
to the driver initialization function. Simplify the routines that check
the requested period and find a suitable channel for driving the pin.

**drivers: pwm_nrf5_sw: Make the driver support nRF9x series SoCs as well**

Now the driver uses DPPI or PPI depending on the presence of one of
these interfaces in a given SoC, hence it can be used on nRF9x SoCs
as well. Remove the '5' from its name then, and make the blink_led
sampe use it for the nrf9160_pca10090 board, thus making the sample
working properly on this board.

Fixes #15323.

anangl committed:
drivers: pwm_nrf5_sw: Use dynamically adjusted timer prescaler

Adjust the timer prescaler used by the pwm_nrf5_sw driver according to
the requested period length, so that it does not need to be specified
at the build time. Remove the incorrect "scaling" that was done to the
requested pulse and period cycles to fit them into 16 bits.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
samples: blink_led: Use nRF SW PWM driver for nRF52 boards

The nrfx Hardware PWM driver cannot be used in this sample
since the driver does not allow to set a period longer than ~262 ms.

The sw_pwm node is made enabled by default in dts for all nRF52 series
SoCs (just like it is already done for nRF51 series) so that there is
no need to do it for particular boards to make the SW PWM driver
available for them (this will only cause that the corresponding DT_
macros will be created, the driver will be compiled and used after
it is selected in Kconfig).

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: pwm_nrf5_sw: Correct the use of GPIOTE channels and TIMER

GPIOTE base channel number taken from DeviceTree was not used
consistenly, hence the driver could properly work only when this
number had the value of 0.
TIMER was not stopped during the reconfiguration of the channel
and this could lead to the inversion of the PWM polarity.
The config structure was removed on the occasion as it unnecessarily
complicated the code of this currently single-instance only driver.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: pwm_nrf5_sw: Use nrfx HALs instead of direct register accesses

Now instead of accessing peripheral registers directly, the driver
uses the nrfx HAL functions. This makes the code easier to maintain
and also gives the possibility to use pins from GPIO port P1 on
nRF52840 (absolute numbers 32-47) as the GPIO HAL provides the needed
translation.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: pwm_nrf5_sw: Refactor the code

Move all channel initialization code that does not need to be executed
on every pin reconfiguration (like initialization of the PPI channels)
to the driver initialization function. Simplify the routine that checks
the requested period and the one that finds a suitable channel for
driving the pin.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: pwm_nrf5_sw: Make the driver support nRF9x series SoCs as well

Now the driver uses DPPI or PPI depending on the presence of one of
these interfaces in a given SoC, hence it can be used on nRF9x SoCs
as well. Remove the '5' from its name then, and make the blink_led
sample using it for the nrf9160_pca10090 board, thus making the sample
working properly on this board.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
Fixed checkpatch and codeowners issues.

ioannisg commented:
Is there a filed bug for this PR? Otherwise, it shouldn't go to 1.14 - we are close to release, AFAIK

anangl commented:
> Is there a filed bug for this PR? Otherwise, it shouldn't go to 1.14 - we are close to release, AFAIK

I am not aware of any. The bug was reported on slack:
```
Aaron Tsui 9:15 AM, March 22nd
BTW, As the PWM_CLK lowest is 125khz(16M/128) and COUNTERTOP max is 0x7fff on nrf52, my understanding the nrf52 HW PWM longest period is about 262ms, cannot reach period as 1000ms. right?
@gaute this will cause failure on sample/blink_led, which need a 1000ms period.
```

anangl commented:
cc @overheat

overheat commented:
@anangl @ioannisg I didn't raise an issue specific for this topic. But I think this is related to #14749, which is verifying samples. Since this is a basic sample bug, I think it's better to be solved before the LTS version released.

anangl commented:
Added one line that I somehow lost during splitting the changes into several logical commits. Sorry about that.

anangl commented:
@ioannisg 
> > Is there a filed bug for this PR? Otherwise, it shouldn't go to 1.14 - we are close to release, AFAIK
> 
> I am not aware of any. The bug was reported on slack:

[..] 

> Sure. Could you file an issue in GH with "bug" and 1.14" that relates to the slack report?

Created #15323.

anangl commented:
Removed a no longer needed overlay file that was changing the `clock-prescaler` property eliminated by this PR.

MaureenHelm commented on CODEOWNERS:
```diff
@@ -218,7 +218,6 @@
 /lib/posix/                               @pfalcon
 /kernel/device.c                          @andrewboie @andyross @nashif
 /kernel/idle.c                            @andrewboie @andyross @nashif
-/samples/basic/servo_motor/*microbit*     @jhe
```
Looks like a stray change that doesn't belong in this commit

anangl replied:
I have split the commit into two, to make the removal of the "clock-prescaler" property and the above entry from CODEOWNERS more visible and clear.

MaureenHelm commented on drivers/pwm/pwm_nrf5_sw.c:
```diff
@@ -208,12 +230,11 @@ static int pwm_nrf5_sw_pin_set(struct device *dev, u32_t pwm,
-	*cycles = 16000000UL / BIT(config->prescaler);
+	/* Since this function might be removed (see issue #6958), the maximum
+	 * supported frequency (16 MHz) is always returned here, and dynamically
+	 * adjusted timer prescalers are used in the pin set function.
```
Not sure I understand this logic. Might the application want to know the actual frequency after scaling, perhaps to know the current timing range/resolution?

anangl replied:
We use the same approach in [pwm_nrfx](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/pwm/pwm_nrfx.c#L236). It is much more convenient for the users to have the prescaler computed dynamically (see the discussion in #5469 and the referred #6958) instead of having to set it at the build time.

MaureenHelm commented on dts/arm/nordic/nrf52810.dtsi:
```diff
@@ -200,7 +200,7 @@
-	clock-prescaler = <0>;
 	ppi-base = <14>;
 	gpiote-base = <0>;
+	status = "ok";
```
Better to do this at the board level instead.

anangl replied:
Actually, I'm not convinced it would be better. The activation of the nRF SW PWM driver is done through Kconfig option, so this enabling here leads only to generation of `DT_` macros, which is harmless. And it's more convenient to enable the node here than in all boards that uses this SoC. Am I missing some important consequence of doing it this way?

MaureenHelm commented on samples/basic/blink_led/boards/nrf52810_pca10040.conf:
```diff
@@ -0,0 +1,3 @@
+# The nrfx Hardware PWM driver cannot be used in this sample since the driver
+# does not allow to set a period longer than ~262 ms.
+CONFIG_PWM_NRFX=n
```
Can you add to this comment that you're disabling the HW PWM driver in order to enable the SW PWM driver?

anangl replied:
Done.

MaureenHelm commented on drivers/pwm/pwm_nrf5_sw.c:
```diff
@@ -14,41 +14,34 @@
 #define MAX_TIMER_VALUE  (GET_TIMER_CAPABILITY(MAX_SIZE) == 32 ? UINT32_MAX \
 							       : UINT16_MAX)
 
-struct pwm_config {
```
Can the hardware support multiple instances? If so, this is a step in the wrong direction. Why not fix the driver to use the device tree data consistently?

anangl replied:
The driver actually combines several peripherals to achieve the PWM functionality. But yes, it can be made multi-instance (though, due to the checks of capabilities of the selected TIMER, this will probably require more macro magic). I was not sure if it would be desirable and didn't want to do this in this PR. The removal of this structure was to clean up the current state (the driver was single-instance from its origin).

anangl committed:
dts: bindings: nordic,nrf-sw-pwm: Remove "clock-prescaler" property

Remove the no longer used "clock-prescaler" property from the bindings
for "nordic,nrf-sw-pwm" compatible nodes and all assignments to this
property. This means also the removal of the bbc_microbit.overlay file
from the servo_motor sample (as the assignment is the only purpose
of this file) together with the entry covering this file in CODEOWNERS.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

henrikbrixandersen commented:
Do these change also address the issue raised in #8628 (Fix polarity issue when using multiple channels) in the SW PWM driver?

anangl commented:
@henrikbrixandersen 
> Do these change also address the issue raised in #8628 (Fix polarity issue when using multiple channels) in the SW PWM driver?

Yes, they do. By [stopping the timer](https://github.com/zephyrproject-rtos/zephyr/pull/15287/files#diff-bc88d4b3519c4c50a750394fa1c95146R172). But I think I'll try to use the solution with PPI channel group, as proposed by @cvinayak in https://github.com/zephyrproject-rtos/zephyr/pull/8628#discussion_r210176762.

anangl commented:
Applied a couple of minor corrections.

anangl commented:
Rebased on master.

anangl committed:
drivers: pwm_nrf_sw: Allocate PPIs instead of assigning them statically

This approach provides more flexibility (the PPI channels utilized by
this driver no longer need to have consecutive numbers) and eliminates
the need of certain hacks in nrfx_glue.h and radio_nrf5_ppi.h.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
Applied adjustments needed after rebasing.

anangl commented:
Rebased on master to resolve the conflict in `samples/basic/servo_motor/src/main.c`.

carlescufi commented:
@MaureenHelm can you please take a fresh look at this one?

carlescufi commented:
@cvinayak @kl-cruz @nordic-krch can you please take a look?

nordic-krch commented on drivers/pwm/pwm_nrf_sw.c:
```diff
@@ -0,0 +1,303 @@
+
+#define LOG_LEVEL CONFIG_PWM_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(pwm_nrf_sw);
```
use `LOG_MODULE_REGISTER(pwm_nrf_sw, CONFIG_PWM_LOG_LEVEL);`

anangl replied:
done

nordic-krch commented on drivers/pwm/pwm_nrf_sw.c:
```diff
@@ -0,0 +1,303 @@
+							       : UINT16_MAX)
+
+#if (GET_TIMER_CAPABILITY(CC_NUM) == 6)
+#define PWM_PERIOD_TIMER_CHANNEL  5
```
as i don't like ifdefs i would do `#define PWM_PERIOD_TIMER_CHANNEL ((GET_TIMER_CAPABILITY(CC_NUM) == 6) ? 5 : 3)`

anangl replied:
done

anangl commented:
Rebased on master.

nordic-krch commented:
@anangl is there any way to detect if there is conflict in TIMER instance usage?

anangl commented:
Replaced (as a fix for #15323) by #20369.
